### DISCUSSION TITLE
How to use "Azure Translator API Action Button" by Nobeard

### DISCUSSION THREAD

**Discussion by Hamsterpants** [thetechcollective/public-discussions#14] [Open Source]
## Prerequisites
- **Microsoft Azure Subscription**
- **Azure DevOps Account**
- **Rights to Create Custom Processes** in an Organization on Azure DevOps
- **Rights to Download Extensions** in an Organization on Azure DevOps

## Microsoft Azure AI services
1. **Create Azure AI Translator Service**
	- Go to start page of your Microsoft Azure `portal.azure.com`.
	- Click `Create a resource` .
	- Under `Categories > AI + Machine Learning`, search for `Translator` and click on it.
    - Configure the resource as needed.
2. **Get Key and Region**
	- Go to the newly created resource.
	- Under `Resource Management` click `Keys and Endpoints`.
    - Copy either Key 1 or Key 2 for use in the extension.
    - You can also get Location/Region from here, e.g., `swedencentral`.
### Download extension

1. **Navigate to Extensions Settings**
    - Go to the start page of your organization: `dev.azure.com/ORGNAME/`.
    - Go to `Organization Settings` at the bottom left.
    - Go to `Extensions` under `General`.

2. **Install Shared Extension**
    - Under the `Shared` tab, you will see a view of shared extensions.
    - Click the extension and then click `Install` in the top right.

3. **Install Public Extension**
    - Navigate to `marketplace.visualstudio.com`.
    - Search for the extension and install it.
    - Choose which organization to download it for.
    - It should now appear under the `Installed` tab.

### Create a Custom Process Template

1. **Navigate to Process Settings**
	- Go to the start page of your organization: `dev.azure.com/ORGNAME/`.
	- Go to `Organization Settings` at the bottom left.
	- Go to `Process` under `Boards`.
    - Click the "•••" next to `Basic (default)` and select `Create inherited process`.

3. **Edit the Created Process**
    - Click on the newly created process.
    - Select the work item type to edit, e.g., `Issue`.

4. **Add a Custom Field**
    - Click `New field` in the top bar.
    - Name the field for reference in the extension.
    - Ensure the field type is `Text (multiple lines)`.
    - The reference name can be found under `••• > Edit > Options` Top right when mouse is over field.

5. **Add a Custom Control**
    - In the top bar, select `Add custom control`.
    - Configure the custom control:
        - **Definition**: Select `Azure Translator API Action Button (NoBeard)`.
        - **Options**:
            - Text to display: `Translate to German`
            - Source text field location: `System.Description`
            - Target text field for translation: `Custom.fieldName`
            - API key from `Azure AI services | Translator`
            - Region of API: `swedencentral`
            - Language of source text: `en`
            - Language of translated text: `de`
    - Press OK.
    - The custom control should now be under `Planning`.
 ![Pasted image 20240423155227](https://github.com/thetechcollective/public-discussions/assets/61192057/220f8709-c6c9-4e3a-a429-704512b1cff8)
### Apply Custom Process Template to a Project

1. **Change Project Process**
    - Go to `Organization Settings > Boards > Process`.
    - Click the number under `Team projects` in `Basic (default)`.
    - Click `•••` and select `Change process`.
    - Choose the created custom process.
    - The project should now be using the custom process.

Issues in the project should now look like this:
![Pasted image 20240423155851](https://github.com/thetechcollective/public-discussions/assets/61192057/64983a4e-ade7-4e1c-8edd-c9cad56a489d)
Clicking the translate button will give:
![image](https://github.com/thetechcollective/public-discussions/assets/61192057/3322d48d-c45c-4415-9102-fb2f00d49e33)

[Link to extension](https://marketplace.visualstudio.com/items?itemName=NoBeard.azure-translator-api-extension&ssr=false#overview)

**Comment by Hamsterpants** [comment]
@falkenberg79

**Comment by falkenberg79** [comment]
@Hamsterpants jeg kan ikke se at der er sket den stoire forskel i TTC Azure DevOps organisationen? Bør jeg kunne gå ind på et vilkårligt work item i et projekt i TTC ADO organisationen nu, og se fanen "Translate" og dermed oversætte til andet sprog?

**Comment by falkenberg79** [comment]
@Hamsterpants Hvis man sætter den op på en test case, så kommer der en masse "snask" med ind i oversættelsen. Det er uden tvivl pga. måden feltet Microsoft.VSTS.TCM.Steps er bygget op på sammenlignet med feltet System.Description. Kan vi gøre noget ved det?
![translation extension on test cases](https://github.com/thetechcollective/public-discussions/assets/122284020/06dcb5d5-d9f1-471e-b817-6a63fc809b99)

**Reply by falkenberg79** [reply]
Derudover kan den også kun oversætte indholdet fra det første step.

**Reply by Hamsterpants** [reply]
Jeg er ikke erfaren med test cases i azure devops, men tror helt sikkert der er en løsning / "hack"

**Comment by falkenberg79** [comment]
En rigtig god og udførlig vejledning synes jeg :)

- Uden at have forsøgt at gennemføre vejledningen (ved jeg ikke om jeg kan), så har jeg nogle få kommentarer nedenfor, som du kan forholde dig til @Hamsterpants. 
- Derudover tænker jeg om der er endnu en Prerequisite at nævne, nemlig hvilke rettigheder/licens ens bruger skal besidde i ADO, for at kunne gennemføre vejledningen.
- Bør vi nævne i vejledningen, at man også kan vælge, at den oversatte tekst kan placeres under fanen "Summary" på en test case hvilket gør, at man kan se den oversatte tekst, når man "runner" sin test case i en browser?

-----

### Microsoft Azure AI services
1. Create Azure AI Translator Service
o Configure the resource as needed 
_**(ved man hvad "as needed" betyder? Her mangler måske også et step om, at man skal klikke på "Create", før man kan konfigurere den?)**_

2. Get Key and Region
o Copy either Key 1 or Key 2 for use in the extension. 
_**(ved man hvilken en man skal vælge?)**_

### Azure DevOps
Download extension
_**Giver det sig selv, hvorfor man skal installere både en Shared og en Public extension?**_

**Reply by Hamsterpants** [reply]
Jeg smider lidt flere prerequisites på.
Guiden var tænkt som et simpelt "quickstart" og det med summary lyder som noget, der er mere skræddersyet.

---

### Angående Microsoft Azure

- **Generel Bemærkning**: Med mindre det er første gang man er derinde, så burde det være meget sigende.
- **Oprettelse af Resource**: Jeg kiggede lige selv, og der er flere måder at oprette en Resource på. Nogle gange skulle jeg klikke "create" og andre gange ikke, så jeg valgte bare den rute, som var mest ligetil.

---

### Om Key1 og Key2

- **Valg af Key**: Om man bruger Key1 eller Key2 er ligegyldigt.
- **Formål**: Det er tænkt som, at man kan give brugeren den Key, de ikke bruger, så man kan regenerere den gamle key uden at de mister funktionaliteten.

---

### Download af Extension

Jeg har gjort det tydeligt, at man enten kan downloade en shared extension, hvis den er privat, eller finde den public.

> **Bemærk**: Det skal nok kun være public i sidste version, da det med shared er mere tænkt til Christines Usecase.

**Comment by falkenberg79** [comment]
[like]  Rasmus Falkenberg - The Tech C... reacted to your message:
________________________________

**Comment by assofohdz** [comment]
@Hamsterpants where is the code hosted for this plugin? I just want to transfer this discussion to that repo

**Reply by Hamsterpants** [reply]
Uhmm it was hosted here: https://github.com/thetechcollective/translate-ext, but seems to be deleted? 
Not sure what happened, maybe you have access to some insights here https://github.com/organizations/thetechcollective/settings/deleted_repositories?

**Reply by assofohdz** [reply]
I have granted our tech-that team in thetechcollective org read-access to all repos. It should show now - can you try agani ?

**Reply by Hamsterpants** [reply]
The unmerged code is currently located on the `bugfix/kevinwh/fix-steps-translation` branch. It was originally developed under private access with the intention of eventually submitting a PR from a fork, but that workflow was never completed. With NoBeard recently launching a new open-source organization to host the extension, it is unclear exactly how he expects this existing work to be migrated or handled.
